sdk/attestation/azure-security-attestation/src/attestation_client.cpp

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/attestation/attestation_client.hpp"
+#include <shared_mutex>
+
+#include "private/attestation_client_models_private.hpp"
+#include "private/attestation_common_request.hpp"
+#include "private/attestation_deserializer.hpp"
+#include "private/package_version.hpp"
+#include <azure/core/base64.hpp>
+#include <azure/core/http/policies/policy.hpp>
+#include <azure/core/internal/http/pipeline.hpp>
+
+#include <string>
+
+using namespace Azure::Security::Attestation;
+using namespace Azure::Security::Attestation::_detail;
+using namespace Azure::Security::Attestation::Models::_detail;
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::_internal;
+
+std::string AttestationClient::ClientVersion() const { return PackageVersion::ToString(); }
+
+AttestationClient::AttestationClient(
+    std::string const& endpoint,
+    std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+    AttestationClientOptions options)
+    : m_endpoint(endpoint)
+{
+  std::vector<std::unique_ptr<HttpPolicy>> perRetrypolicies;
+  if (credential)
+  {
+    //      Azure::Core::Credentials::TokenRequestContext const tokenContext
+    //          = {{_internal::UrlScope::GetScopeFromUrl(m_vaultUrl)}};
+
+    //      perRetrypolicies.emplace_back(
+    //          std::make_unique<BearerTokenAuthenticationPolicy>(credential,
+    //          std::move(tokenContext)));
+  }
+  m_apiVersion = options.Version.ToString();
+  std::vector<std::unique_ptr<HttpPolicy>> perCallpolicies;
+
+  m_pipeline = std::make_shared<Azure::Core::Http::_internal::HttpPipeline>(
+      options,
+      "Attestation",
+      PackageVersion::ToString(),
+      std::move(perRetrypolicies),
+      std::move(perCallpolicies));
+}
+
+Azure::Response<AttestationOpenIdMetadata> AttestationClient::GetOpenIdMetadata(
+    Azure::Core::Context const& context)
+{
+  auto request = AttestationCommonRequest::CreateRequest(
+      m_endpoint, HttpMethod::Get, {".well-known/openid-configuration"}, nullptr);
+
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+  auto openIdMetadata(OpenIdMetadataDeserializer::Deserialize(response));
+  return Response<AttestationOpenIdMetadata>(std::move(openIdMetadata), std::move(response));
+}
+
+Azure::Response<std::vector<AttestationSigner>>
+AttestationClient::GetAttestationSigningCertificates(Azure::Core::Context const& context)
+{
+  auto request
+      = AttestationCommonRequest::CreateRequest(m_endpoint, HttpMethod::Get, {"certs"}, nullptr);
+
+  auto response = AttestationCommonRequest::SendRequest(*m_pipeline, request, context);
+  auto jsonWebKeySet(JsonWebKeySetSerializer::Deserialize(response));
```

Comment:
It is good for readability to know that is going to be the type.  `auto` is amazing, but in cases like this, hides the type because we don't know what Deserialize returns.

it's like `auto something(GetSomething())` v/s `std::String something(GetSomething());`

Reply:
FooSerializer::Serialize take a foo and returns a string - always. FooSerializer::Deserialize take a response or json and returns a Foo - always.

This is a convention applied to all my model serializer types.

OTOH, the SendRequest response type should be called out explicitly.

Reply:
I won't make a big deal about it 😄

Reply:
FWIW, I'll make this more explicit in the documentation for the serializer methods.